### DISCUSSION TITLE
Would mutation observing help to wait for elements even more reliably?

### DISCUSSION THREAD

**Discussion by mrts** [selenide/selenide#2083] [Ideas]
First and foremost, thank you for Selenide, it is really awesome! However, even though it has smart waiting, timeouts are sometimes still an issue.

The [`MutationObserver`](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) interface allows you to observe changes to the DOM and react to them. I recently used it in JavaScript to wait for elements to load and it was a pleasure to work with.

What do you think, in case Selenium would expose the mutation observing API, would that make Selenide even more reliable or not?

This is just a wild idea, currently the mutation observing API is not exposed - it runs in the context of the page that you are interacting with, so you will not have direct access to the JavaScript observer object from your Selenium/Selenide code. And I'm not sure if exposing this API to Selenium is even possible.

**Comment by asolntsev** [comment]
@mrts Yes, sounds interesting. Yes, it belongs to Selenide priorities to make tests more stable by using whatever technical tools. MutationObserver seems one of such tools.

**Comment by mrts** [comment]
For reference, here's a code snippet that demonstrates how to wait for an element using the `MutationObserver`:

```js
function waitForElement(selector) {
  return new Promise(resolve => {
    if (document.querySelector(selector)) {
      console.log(`waitForElement: found ${selector} immediately`);
      return resolve(document.querySelector(selector));
    }

    const observer = new MutationObserver(mutations => {
      if (document.querySelector(selector)) {
        console.log(`waitForElement: found ${selector} after DOM changed`);
        resolve(document.querySelector(selector));
        observer.disconnect();
      }
    });
    observer.observe(document.body, {
      childList: true,
      subtree: true
    });
  });
}

waitForElement(".hello-world").then(element => console.log(`${element} is now present in the DOM`));
```
